src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
 				immutable = false;
 			}
 
-			var stuff = GetStuffToCheckForMember(
-				type,
-				typeSyntax,
-				initializer
-			);
+			// Check all initializations of the member for immutability
+			foreach( ExpressionSyntax initializer in initializers ) {
+				var stuff = GetStuffToCheckForMember(
+					type,
+					typeSyntax,
+					initializer
+				);
 
-			if ( stuff == null ) {
-				// null is a signal that there is nothing further that needs to
-				// be checked
-				return immutable;
-			}
+				if( stuff == null ) {
+					// Null is a signal that there is nothing further that needs to
+					// be checked
+					continue;
+				}
 
-			var (typeToCheck, checkKind, getLocation) = stuff.Value;
+				var (typeToCheck, checkKind, getLocation) = stuff.Value;
+
+				if( m_context.IsImmutable(
+					typeToCheck,
+					checkKind,
+					getLocation,
+					out var diagnostic
+				) ) {
+					continue;
+				}
 
-			if( !m_context.IsImmutable( typeToCheck, checkKind, getLocation, out var diagnostic ) ) {
 				diagnosticSink( diagnostic );
 				immutable = false;
 			}
 
 			return immutable;
 		}
 
+
+		/// <summary>
+		/// A readonly field or property may be re-initialized in constructors.
+		/// This method finds all initializations of the field/property.
+		/// </summary>
+		/// <param name="symbol">The field/property to search for</param>
+		/// <param name="initialInitializer">The initialization with the declaration</param>
+		/// <returns>An immutable list of initializations</returns>
+		private ImmutableArray<ExpressionSyntax> GetInitializers(
+			ISymbol symbol,
+			ExpressionSyntax initialInitializer
+		) {
+			List<ExpressionSyntax> initializers = new() {initialInitializer};
+
+			// If the containing symbol isn't a named type (class) then there
+			// will be no constructors
+			if( symbol.ContainingSymbol is not INamedTypeSymbol namedTypeSymbol ) {
+				return initializers.ToImmutableArray();
+			}
+
+			// Iterate through all constructors in the class
+			foreach( IMethodSymbol methodSymbol in namedTypeSymbol.Constructors ) {
+				// Make sure the syntax of the constructor exists
+				if( methodSymbol.DeclaringSyntaxReferences.Length == 0 ) {
+					continue;
+				}
+
+				// Retrieve the constructor syntax
+				var methodSyntax = methodSymbol.DeclaringSyntaxReferences
+				                               .Single()
+				                               .GetSyntax() as ConstructorDeclarationSyntax;
+
+				// Make sure the constructor actually contains a body

```

Comment:
Are there constructors without a body?

Reply:
You wouldn't expect so, but the IDE was warning me of potential `NullReferenceException`s so I figured it's best to listen to it and be safe.

Reply:
Hm, I wouldn't listen to the IDE then.

The problem with quietly backing away like this in the analyzer when we find unexpected syntax is it creates room for bugs. The code here is saying "well, if it doesn't have a body I guess stuff is fine.". What if there is a form of constructor that doesn't have a body but does write to fields? It doesn't seem likely to me, but I would rather we crashed with a NullReferenceException :)

Reply:
In cases where we are less sure we have made diagnostics, e.g.: https://github.com/Brightspace/D2L.CodeStyle/blob/master/src/D2L.CodeStyle.Analyzers/Diagnostics.cs#L473-L481

This one makes sense because C# may add a new type kind in the future.

For this case, it seems unlikely to me that there is going to be new syntax for bodyless constructors (I could be wrong, maybe it exists now!) so I'm happy with a NullReferenceException (its simpler.)

Reply:
I believe it's `Body` vs `ExpressionBody` (arrow syntax). They're mutually exclusive and the other is null.

Reply:
So that is to say, we need to handle `ExpressionBody`.

Reply:
Are expression bodies supported for constructors?

Reply:
Oh damn they are.

Aside: we should have test cases for the destructuring assignment, that might be its own case I'm not sure. Probably bettter as a follow-up PR though.

Reply:
`foo.Body ?? foo.ExpressionBody`

Reply:
> Oh damn they are

For specific example: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members#constructors